File: src/test/java/info/szadkowski/contact/controller/MessageControllerTest.java
package info.szadkowski.contact.controller;

import info.szadkowski.contact.controller.exception.ExceptionHandlerController;
import info.szadkowski.contact.model.MessageRequest;
import info.szadkowski.contact.throttle.Throttler;
import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Test;
import org.junit.jupiter.api.extension.ExtendWith;
import org.mockito.Mock;
import org.mockito.Mockito;
import org.mockito.junit.jupiter.MockitoExtension;
import org.springframework.http.MediaType;
import org.springframework.test.web.servlet.MockMvc;
import org.springframework.test.web.servlet.setup.MockMvcBuilders;

import java.util.ArrayList;
import java.util.List;

import static org.assertj.core.api.Assertions.assertThat;
import static org.springframework.test.web.servlet.request.MockMvcRequestBuilders.post;
import static org.springframework.test.web.servlet.result.MockMvcResultMatchers.status;

@ExtendWith(MockitoExtension.class)
class MessageControllerTest {
  private List<MessageRequest> messages;

  @Mock
  private Throttler throttler;

  private MockMvc mvc;

  @BeforeEach
  void setUp() {
    messages = new ArrayList<>();
    Mockito.when(throttler.canProcess(Mockito.anyString())).thenReturn(true);
    MessageController messageController = new MessageController(
            messages::add,
            r -> "templated:" + r.get("content"),
            throttler,
            throttler);
    mvc = MockMvcBuilders.standaloneSetup(messageController)
            .setControllerAdvice(new ExceptionHandlerController())
            .build();
  }

  @Test
  void shouldSendCorrectMailRequest() throws Exception {
    mvc.perform(post("/v1/message")
            .contentType(MediaType.APPLICATION_JSON_UTF8)
            .content("" +
                    "{" +
                    "\"subject\": \"mySubject\"," +
                    "\"content\": \"myContent\"" +
                    "}")
    ).andExpect(status().isOk());

    assertThat(messages).containsExactly(MessageRequest.builder()
            .subject("mySubject")
            .content("templated:myContent")
            .build());
  }
}

File: src/test/kotlin/info/szadkowski/contact/controller/MessageControllerTest.kt
package info.szadkowski.contact.controller

import info.szadkowski.contact.controller.exception.ExceptionHandlerController
import info.szadkowski.contact.model.MessageRequest
import info.szadkowski.contact.service.MessageService
import info.szadkowski.contact.throttle.Throttler
import org.junit.jupiter.api.BeforeEach
import org.junit.jupiter.api.Test
import org.springframework.http.MediaType
import org.springframework.test.web.reactive.server.WebTestClient
import strikt.api.expectThat
import strikt.assertions.containsExactly

class MessageControllerTest : MessageService {
    lateinit var messageService: (MessageRequest) -> Unit
    lateinit var client: WebTestClient

    override suspend fun send(message: MessageRequest) = messageService(message)

    @BeforeEach
    fun setUp() {
        val dummyThrottler = object : Throttler {
            override fun canProcess(key: String) = true
            override fun clearExpired() = Unit
        }
        val messageController = MessageController(this, dummyThrottler, dummyThrottler) {
            "templated:" + this["content"]
        }
        client = WebTestClient
            .bindToController(messageController)
            .controllerAdvice(ExceptionHandlerController())
            .webFilter<WebTestClient.ControllerSpec>(IpReadingFilter())
            .build()
    }

    @Test
    fun `Should send correct mail request`() {
        val messages = mutableListOf<MessageRequest>()
        messageService = { messages.add(it) }

        client.post()
            .uri("/v1/message")
            .contentType(MediaType.APPLICATION_JSON)
            .bodyValue(
                """
                    {
                        "subject": "mySubject",
                        "content": "myContent"
                    }
                """.trimIndent()
            )
            .exchange()
            .expectStatus().isOk

        expectThat(messages).containsExactly(
            MessageRequest(
                subject = "mySubject",
                content = "templated:myContent"
            )
        )
    }

    @Test
    fun `Should fail on incorrect mail request`() {
        messageService = { throw MessageService.MessageSendException() }

        client.post()
            .uri("/v1/message")
            .contentType(MediaType.APPLICATION_JSON)
            .bodyValue("{}")
            .exchange()
            .expectStatus().isBadRequest
    }
}
